src/canthrow.d

```diff
@@ -64,50 +65,124 @@ extern (C++) bool canThrow(Expression e, FuncDeclaration func, bool mustNotThrow
+                return;
+
+            if (mustNotThrow)
             {
-            }
-            else if (t.ty == Tfunction && (cast(TypeFunction)t).isnothrow)
-            {
-            }
-            else if (t.ty == Tdelegate && (cast(TypeFunction)(cast(TypeDelegate)t).next).isnothrow)
-            {
-            }
-            else
-            {
-                if (mustNotThrow)
+                if (ce.f)
                 {
-                    const(char)* s;
-                    if (ce.f)
-                        s = ce.f.toPrettyChars();
-                    else if (ce.e1.op == TOKstar)
-                    {
-                        // print 'fp' if ce->e1 is (*fp)
-                        s = (cast(PtrExp)ce.e1).e1.toChars();
-                    }
-                    else
-                        s = ce.e1.toChars();
-                    ce.error("'%s' is not nothrow", s);
+                    ce.error("%s '%s' is not nothrow",
+                        ce.f.kind(), ce.f.toPrettyChars());
+                }
+                else
+                {
+                    auto e1 = ce.e1;
+                    if (e1.op == TOKstar)   // print 'fp' if e1 is (*fp)
+                        e1 = (cast(PtrExp)e1).e1;
+                    ce.error("'%s' is not nothrow", e1.toChars());
                 }
-                stop = true;
             }
+            stop = true;
         }
 
         override void visit(NewExp ne)
         {
             if (ne.member)
             {
+                if (ne.allocator)
+                {
+                    // Bugzilla 14407
+                    Type t = ne.allocator.type.toBasetype();
+                    if (t.ty == Tfunction && !(cast(TypeFunction)t).isnothrow)
+                    {
+                        if (mustNotThrow)
+                        {
+                            ne.error("%s '%s' is not nothrow",
+                                ne.allocator.kind(), ne.allocator.toPrettyChars());
+                        }
+                        stop = true;
+                    }
+                }
                 // See if constructor call can throw
                 Type t = ne.member.type.toBasetype();
                 if (t.ty == Tfunction && !(cast(TypeFunction)t).isnothrow)
                 {
                     if (mustNotThrow)
-                        ne.error("constructor %s is not nothrow", ne.member.toChars());
+                    {
+                        ne.error("%s '%s' is not nothrow",
+                            ne.member.kind(), ne.member.toPrettyChars());
+                    }
                     stop = true;
                 }
             }
             // regard storage allocation failures as not recoverable
         }
 
+        override void visit(DeleteExp de)
+        {
+            Type tb = de.e1.type.toBasetype();
+            AggregateDeclaration ad = null;
+            switch (tb.ty)
+            {
+                case Tclass:
```

Comment:
Sigh. This PR has opened in cdmd age. ddmd conversion had introduced many style breaks...
Fixed.